### PR TITLE
Tests for DateUtils and FileExtensions

### PR DESCRIPTION
**Pull request by sgallese**
Test coverage for FileExtensions and DateUtils
Prepare classes for migration in #1075

### CODE REVIEW

**Comment by sgallese** [comment]
@iSoron  These tests pass, but I am a bit confused by the DateUtil implementations for:
getStartOfTodayCalendarWithOffset() ([test1](https://github.com/iSoron/uhabits/pull/1111/files#diff-364b126afb4194c42ba8ce5331c76a407f1cfaeb137327c5f1dc757043ffa29bR313 ), [test 2](https://github.com/iSoron/uhabits/pull/1111/files#diff-364b126afb4194c42ba8ce5331c76a407f1cfaeb137327c5f1dc757043ffa29bR327 )) 
getStartOfTodayWithOffset() ([test1](https://github.com/iSoron/uhabits/pull/1111/files#diff-364b126afb4194c42ba8ce5331c76a407f1cfaeb137327c5f1dc757043ffa29bR170 ), [test2](https://github.com/iSoron/uhabits/pull/1111/files#diff-364b126afb4194c42ba8ce5331c76a407f1cfaeb137327c5f1dc757043ffa29bR182)) 

For the above methods, these implementations are returning a non-offset time/calendar.
For example, if the offset is 3 hours, they still return a Midnight time/calendar.

This differs from the behavior of getStartOfTomorrowWithOffset() ([test 1](https://github.com/iSoron/uhabits/pull/1111/files#diff-364b126afb4194c42ba8ce5331c76a407f1cfaeb137327c5f1dc757043ffa29bR146), [test 2](https://github.com/iSoron/uhabits/pull/1111/files#diff-364b126afb4194c42ba8ce5331c76a407f1cfaeb137327c5f1dc757043ffa29bR158))
For example, if the offset is 3 hours, they return a 3AM time/calendar.

If this is not expected, let me know the correct expectation and I can create an issue and resolve.

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt [hunk 1]
```diff
@@ -166,6 +315,45 @@ class DateUtilsTest : BaseUnitTest() {
...
+        setStartDayOffset(hourOffset, 0)
+        val afterOffset = unixTime(2017, Calendar.JANUARY, 1, hourOffset + 1, 0)
+        setFixedLocalTime(afterOffset)
+        setFixedLocale(Locale.GERMANY)
+        val startOfToday = unixTime(2017, Calendar.JANUARY, 1, 0, 0)
+        val expectedCalendar = GregorianCalendar(TimeZone.getTimeZone("GMT"), Locale.GERMANY)
+        expectedCalendar.timeInMillis = startOfToday
+        val startOfTodayCalendar = DateUtils.getStartOfTodayCalendar()
+        assertThat(expectedCalendar, equalTo(startOfTodayCalendar))
+    }
```
Comment (iSoron) [inline uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt:355]: Same comment for these two tests.

**Review by iSoron — CHANGES_REQUESTED** [review]
Thank you for the tests, @sgallese. I think they are good. I left just some minor comments below.

**Comment by iSoron** [comment]
> For the above methods, these implementations are returning a non-offset time/calendar.
> For example, if the offset is 3 hours, they still return a Midnight time/calendar.

It's also unclear to me. I suggest that we write tests for the current behavior. Changing it might introduce subtle bugs.

The underlying issue here is that, historically, we have stored all dates as milliseconds since Jan 1, 1970, but, since we don't actually care about hours/minutes/seconds in most situations, we have also implicitly assumed that most millis are a multiple of `DAY_LENGTH`. I suspect that the reason these functions are returning midnight is to keep this assumption valid. We should eventually move everything that doesn't care about hours/minutes/seconds to  `LocalDate`, which solves this ambiguity by storing time at a lower granularity.

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/utils/DateUtils.kt [hunk 2]
```diff
@@ -77,11 +77,11 @@ abstract class DateUtils {
...
         @JvmStatic
-        fun getLocalTime(): Long {
+        fun getLocalTime(testTimeInMillis: Long? = null): Long {
             if (fixedLocalTime != null) return fixedLocalTime as Long
 
             val tz = getTimeZone()
-            val now = Date().time
+            val now = testTimeInMillis ?: Date().time
             return now + tz.getOffset(now)
         }
```
[the remark this answers is not in the document]
Comment (sgallese) [reply uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/utils/DateUtils.kt:86]: Agreed  aae85c

File: uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt [hunk 3]
```diff
@@ -58,6 +59,137 @@ class DateUtilsTest : BaseUnitTest() {
...
+    fun testGetStartOfTomorrowWithOffset_afterOffset() {
+        val hourOffset = 3
+        setStartDayOffset(hourOffset, 0)
+        setFixedTimeZone(TimeZone.getTimeZone("GMT"))
+        val startOfTomorrowWithOffset = unixTime(2017, Calendar.JANUARY, 2, hourOffset, 0)
+        val afterOffset = unixTime(2017, Calendar.JANUARY, 1, hourOffset + 1, 0)
+        setFixedLocalTime(afterOffset)
+        val startOfTomorrow = DateUtils.getStartOfTomorrowWithOffset()
+        assertThat(startOfTomorrowWithOffset, equalTo(startOfTomorrow))
+    }
```
[the remark this answers is not in the document]
Comment (sgallese) [reply uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt]: I was able to move the first three lines to a shared helper function, but the rest of the lines are different. If I'm missing something else obvious for DRYer test code, let me know! 707b2b4

[the remark this answers is not in the document]
Comment (sgallese) [reply uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt]: Actually, ended up writing much DRYer offset tests 4fbf8a8
See if these are better.
Once merged, I'll want to update the new non-JVM tests to reflect this logic in https://github.com/iSoron/uhabits/pull/1120/

**Comment by sgallese** [comment]
@iSoron OK to merge this PR?

**Comment by iSoron** [comment]
@sgallese Sorry, I missed the 'review requested' notification. It looks good now, thanks!

**Pull request merged**